Automattic/jetpack
# Story block: Fix (server-side) rendering of external stories

Currently, we use the media ids to render the story on the server on the server. We even overwrite urls using the data we have for each media. As such, copy pasting a story from one site to another will be an issue as we will be serving different media than the one intended.

To fix that we can rely on the `url` param in each media object stored in the story block attributes then enrich the media with extra data if the corresponding media from the media library matches our url. 

As such we support copy-pasting stories across sites with different media libraries.

We however won't have support for editing those copy pasted stories as the media ids won't match. This is coherent with other blocks behavior such as slideshow.

This PR adds tests for server side rendering of the story block to avoid future regressions on videopress for instance.

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:

* Try rendering a story with images hosted on your site, it should look the same with and without js
* Try rendering a story hosted on your site with a videopress video as the first slide
* Disable js and reload the page, make sure the story looks the same
* Try rendering a story hosted on your site with a non-videopress video as the first slide, with js on it should load the video, without js on it should be loading the next image in the story, if there are none then it will be loading the video anyway, in which case we might see unwanted controls, but that's expected. 
* Try copy pasting a story with images from one site to another. It should work as expected. Copy-paste a story where the first slide media ID already exists on the site, with and without js, it should be using the url inside the story template instead of trying to load the local media ID.
* Try copy pasting a story which first slide is a video from one site to another. Notice that it won't work if you disable js. Even for videos from videopress which have a poster, it is not available on upload, so it's not added to the block's attribute, it's entirely dynamic and as such it will only work on the original site.
* Try it on WPcom
* Try sending yourself an email notification from your wpcom sandbox (After applying this patch to it) using `php ./bin/subscriptions/send.php BLOG_ID post POST_ID EMAIL_ADDRESS`. Try to open the email on different email clients (it's not perfect on GMail for iOS I noticed)
